### PR TITLE
NewTLSConnectionWithOpts

### PR DESCRIPTION
**Pull request by oconnor663**
This is yet another method for creating a Gregor connection. Two goals
here: 1) the mandatory-positional-args-and-optional-args-in-a-struct
pattern is flexible, so hopefully the other interfaces can get merged
into this one over time, and 2) it takes a DialTLS method, so that the
caller can set per-request timeouts.

r? @maxtaco

### CODE REVIEW

File: rpc/connection.go [hunk 1]
```diff
@@ -245,6 +250,40 @@ type Connection struct {
...
+
+func TLSConfigFromRootCerts(rootCerts []byte) (*tls.Config, error) {
+	certs := x509.NewCertPool()
+	if !certs.AppendCertsFromPEM(rootCerts) {
+		return nil, errors.New("Unable to load root certificates")
+	}
+	return &tls.Config{RootCAs: certs}, nil
+}
+
+func NewTLSConnectionWithOpts(
```
Comment (maxtaco) [inline rpc/connection.go]: why not just throw everything into the struct?

File: rpc/connection.go [hunk 2]
```diff
@@ -245,6 +250,40 @@ type Connection struct {
...
+type ConnectionOpts struct {
+	TagsFunc       LogTagsFromContext
+	Protocols      []Protocol
+	DontConnectNow bool
+	LogFactory     LogFactory
+	WrapErrorFunc  WrapErrorFunc
+	DialTLS        func(network, addr string) (net.Conn, error)
+}
+
+func TLSConfigFromRootCerts(rootCerts []byte) (*tls.Config, error) {
```
Comment (maxtaco) [inline rpc/connection.go]: why is this needed?

**Comment by maxtaco** [comment]
@akalin-keybase or @aalness can you take a look too?

**Comment by oconnor663** [comment]
Some more context for why I ended up with `DialTLS` as the API I wanted:

There are two things we want to do with timeouts in general. The first is make sure that the dial and the TLS handshake have timeouts (and maybe other configs, like `KeepAlive`), and the second is to _wrap_ the dialed connection so that it's `Read` and `Write` calls set internal timeouts of their own.

The Go standard libraries are kind of inconsistent about how libraries that make network connections should expose this sort of stuff. Some functions take a `Dialer` object, which makes sense and is pretty clean. The problem there though, is that a `Dialer` is a _struct_ rather than an interface, so all you can do is change its fields -- there's no way to define any kind of wrapper around any of its methods. (Unless I'm really missing something with how Go methods work.)

The next most likely pattern is to let the caller provide a `Dial` or `DialContext` method. The contextified version feels more modern, but unfortunately the `tls` library still isn't context-aware at all (for example, there's no way to cancel the `Handshake`). But even with regular `Dial` there's another issue: The `tls` library is inconsistent between `Client` (which takes an already-dialed connection and does TLS over it) and `DialWithDialer` (which takes a `Dialer` and dials a new connection of its own). `DialWithDialer` allows the `TlsConfig` to be nil and produces some reasonable defaults, while `Client` doesn't allow the config to be nil. If we wanted to take a `Dial` method from our caller, we'd need to call `Client` on it, and that means we'd be responsible for assembling the right `TlsConfig` defaults. This would just be copying some lines from `tls`, so it's totally doable, but it feels pretty gross.

So the final API I've landed on is `DialTLS`. This is taking a page from the `http` library, which allows you to specify either `Dial` or `DialContext` _or_ `DialTLS`. I'd prefer to avoid that kind of API redundancy, so that callers don't have think about what happens when multiple ones are set, but taking a `DialTLS` method by itself seems to allow the caller to solve the wrapping problem and also use the more convenient `DialWithDialer` inside its `DialTLS` implementation.

So yeah...fun times...

**Comment by oconnor663** [comment]
Closing in favor of https://github.com/keybase/client/pull/5110
